### DISCUSSION TITLE
MDX File System Route API doubled directory

### DISCUSSION THREAD

**Discussion by jwz16** [gatsbyjs/gatsby#36291] [Help] [answered]
Hi community, I have been struggling this issue for two days, following **[Create Pages Programmatically](https://www.gatsbyjs.com/docs/tutorial/part-6/)**, any help would be much appreciated.

### Description
When using `{mdx.slug}.js` inside a nested directory in `/pages`, it will generate a route with doubled directory.

### Steps to Reproduce
1. Set up a new Gatsby project with `npm init gatsby test`, selecting "add markdown and MDX support" when going through the questions.
2. Create a `/pages/blog` folder
3. Create any page using MDX under the `/pages/blog/`, e.g. `first-blog.mdx`
4. Run `npm start`.
5. Open browser and type `localhost:8000/bla` and get 404 error.
6. You can see a route named `/blog/blog/first-blog` which is not expected.
...

### Expected Result
Generate `/blog/first-blog` route

### Environment
System:
OS: Linux/Ubuntu 20.04.4
CPU: KVM Processor

Binaries:
Node: v16.16.0
NPM: 8.11.0

**Comment by KyleAMathews** [comment]
It seems the slugs that are created already have blog in them. E.g. what happens if you put the route in a different directory?

So either you can change what slug is created or change where the route is.

**Reply by jwz16** [reply]
Hi Kyle, many thanks for spending your time to help me, I've tried putting the route in a different directory, you can see [the reply below](https://github.com/gatsbyjs/gatsby/discussions/36291#discussioncomment-3297456)

**Comment by LekoArts** [comment, marked as answer]
You're combining two ways of creating pages from MDX and this way the clash. You should choose only one way.

Currently you're doing:

- Automatically create pages from `src/pages` by having e.g. `src/pages/blog/first-blog.mdx`
- Create pages from `src/pages/{mdx.slug}.js`

But since you have MDX files inside `src/pages` you'll create double the routes.

Two options:

- Either delete the {mdx.slug}.js file
- Or move the MDX files outside of `src/pages`, point `gatsby-source-filesystem` to there and create them with {mdx.slug}.js

**Reply by jwz16** [reply]
Hi LekoArts,

Thanks for your suggestions, however I have tried the second option, it gave me the following errors:
```
warn The GraphQL query in the non-page component "/home/ubuntu/projects/experimental/test/src/blog/{mdx.slug}.js" will not be run.
Exported queries are only executed for Page components. It's possible you're
trying to create pages in your gatsby-node.js and that's failing for some
reason.

If the failing component(s) is a regular component and not intended to be a page
component, you generally want to use a <StaticQuery> (https://gatsbyjs.com/docs/static-query)
instead of exporting a page query.

If you're more experienced with GraphQL, you can also export GraphQL
fragments from components and compose the fragments in the Page component
query and pass data down into the child component — https://graphql.org/learn/queries/#fragments
```

My `gatsby-config.js`:
```javascript
module.exports = {
  siteMetadata: {
    title: `Test`,
    siteUrl: `https://www.yourdomain.tld`
  },
  plugins: ["gatsby-plugin-mdx", {
    resolve: 'gatsby-source-filesystem',
    options: {
      "name": "pages",
      "path": "./src/pages/"
    },
    __key: "pages"
  },{
    resolve: 'gatsby-source-filesystem',
    options: {
      "name": "blogs",
      "path": "./src/blog/"
    },
    __key: "blog"
  },]
};
```

**Reply by jwz16** [reply]
Sorry, I may misread the document, after I put the mdx files outside the `pages` directory (say `src/blog`) and leave the `{mdx.slug}.js` under `/pages/blog`, then set the `gatsby-source-filesystem` point to `src/blog`, everything worked as expected right now.

For who may have this issue, my project structure:
```
- src
  - blog
    - first-blog.mdx
  - pages
    - blog
      - {mdx.slug}.js
  - 404.js
  - index.js
```

Thank you all for the help right here!
